File: R/tests/tests/merge_artifacts_propagation.R
test <- new.env()
test$name <- "MergeArtifactsPropagationTest"
test$priority <- 1

test$start <- decorated_function(
  function(self) {
    self$non_modified_passdown <- "a"
  },
  type = "step", prio = 0, qual = c("start"), required = TRUE
)

test$modify_things <- decorated_function(
  function(self) {
    # Set different names to different things
    val <- self$index + 1
    self[[sprintf("val%d", val)]] <- val
  },
  type = "step", prio = 0, qual = c("foreach-inner-small"), required = TRUE
)


test$merge_things <- decorated_function(
  function(self, inputs) {
    merge_artifacts(self, inputs)

    stopifnot(self$non_modified_passdown == "a")
    for (i in 1:length(inputs)) {
      stopifnot(self[[sprintf("val%d", i)]] == i)
    }
  },
  type = "step", prio = 0, qual = c("join"), required = TRUE
)

test$all <- decorated_function(
  function(self) {
    stopifnot(self$non_modified_passdown == "a")
  },
  type = "step", prio = 1, qual = c("all"), required = TRUE
)

File: test/core/tests/merge_artifacts_propagation.py
from metaflow_test import MetaflowTest, ExpectationFailed, steps


class MergeArtifactsPropagationTest(MetaflowTest):
    # This test simply tests whether things set on a single branch will
    # still get propagated down properly. Other merge_artifacts behaviors
    # are tested in the main test (merge_artifacts.py). This test basically
    # only matches with the small-foreach graph whereas the other test is
    # more generic.
    PRIORITY = 1

    @steps(0, ["start"])
    def start(self):
        self.non_modified_passdown = "a"

    @steps(0, ["foreach-inner-small"], required=True)
    def modify_things(self):
        # Set different names to different things
        val = self.index
        setattr(self, "var%d" % (val), val)

    @steps(0, ["foreach-join-small"], required=True)
    def merge_things(self, inputs):
        self.merge_artifacts(
            inputs,
        )

        # Ensure that everything we expect is passed down
        assert_equals(self.non_modified_passdown, "a")
        for i, _ in enumerate(inputs):
            assert_equals(getattr(self, "var%d" % (i)), i)

    @steps(1, ["all"])
    def step_all(self):
        assert_equals(self.non_modified_passdown, "a")
